### ISSUE TITLE
Decision tree classifier porter C code predicting index of classes not actual class

### ISSUE THREAD

**Issue by vijaykilledar** [nok/sklearn-porter#37]
Attaching training data csv file where first column is  target class to predict.
I generated pickle file and using sklearn-porter command line i convert pickle file to C Code and ran it.
C code returning index of classes not actual classes and Python predict() function returns actual class not index.

Attaching  training csv file pickle file.
[csv_and_pickle_file.zip](https://github.com/nok/sklearn-porter/files/2169463/csv_and_pickle_file.zip)

**Comment by nok** [comment]
Yes you are right. In general all estimators in all programming languages return the index of the resulted label (`y`), because it would be an overhead to reimplement the mapping for each programming language. Nevertheless I noted this requirement for a future release.

**Comment by HTCode** [comment]
@nok do you have a hint on how to return the actual label (name) instead of an index ? I am not sure if it's sufficient to add the list of unique label names (given by clf.classes_) to the exported JSON data and then in the target language (I am using Java) do something like unique_label_names[clf.predict(features)] ? But I double if the order in clf.classes_ corresponds to the indexes predicted in Java ...

**Comment by HTCode** [comment]
Or a "not so intelligent" trick would be to call predict on the training samples in the target language, then simply map the predicted indexes to the actual labels of the training set ...

**Comment by skjerns** [comment]
I just came across this problem as well:

In my case, I have input classes ranging from `[1,2,3,4,5]`, however there is no example in the training for class `2`. As a result, the C-version of my random forest outputs classes `[1,2,3,4]`, with `2,3,4` being actually `3,4,5`. Is there any way to prevent that, or are there ideas to fix this?

**Comment by skjerns** [comment]
I solved it temporarily by writing a small wrapper:

It adds a conversion function to the c code and embeds it:

```C
int idx2label(int class_idx) { 
    int labels[5] = {0,2,3,4,5}; // your original ints
    return labels[class_idx];
}
```


```Python
import sklearn
def save_model_sklearn_porter(clf, file):
    porter = sklearn_porter.Porter(clf, language='C')
    output = porter.export(embed_data=True)
       
    # see which labels are in the classifier, so far only ints are supported
    labels = [str(int(i)) for i in clf.classes_]
    # create new label code and conversion function
    labels_code = 'int labels[{}] = {{{}}}'.format(len(labels), ','.join(labels))
    convert_func = '\n\nint idx2label(int class_idx) { \n' +\
                   '    {};\n    return labels[class_idx];\n}}\n\n'.format(labels_code)
    output += convert_func
    
    # replace last occurrence of `return class_idx` with the label transfer function
    # with [::-1] we can revert the string and look for the first element as if it where the last
    output = output[::-1].replace('return class_idx'[::-1], 'return idx2label(class_idx)'[::-1], 1)[::-1]
    
    with open(file, 'w') as file:
        file.write(output)
    return output

```
